### PR TITLE
Change merge return pass to handle structured cfg.

### LINKED ISSUE #857: Add merge-return pass
Transform each  reachable function so that:
- it has at most one basic block with an OpReturn or OpReturnValue
- that basic block post-dominates all reachable blocks (modulo those ending with OpKill)

Expose a general purpose helper function that does this transform on a single SPIR-V function.

Motivated by https://github.com/KhronosGroup/SPIRV-Tools/issues/850

### PR DESCRIPTION
**Pull request by s-perron**
We are seeing shaders that have multiple returns in a functions.  These
functions must get inlined for legalization purposes; however, the
inliner does not know how to inline functions that have multiple
returns.

The solution we will go with it to improve the merge return pass to
handle structured control flow.

Fixes #857.

### CODE REVIEW

**Comment by s-perron** [comment]
This is a first pass.  I think there is still a lot of clean-up, but I would like some feedback sooner rather than later.

The one part still missing is dominators for unreachable loops.  I believe there is a test that exposes that issue.

File: source/opt/merge_return_pass.cpp [hunk 1]
```diff
@@ -15,28 +15,543 @@
...
+    context()->set_instr_block(block->tailInst(), block);
+  } else {
+    block->AddInstruction(MakeUnique<ir::Instruction>(context(), SpvOpReturn));
+  }
+}
+
+void MergeReturnPass::ProcessStructuredBlock(ir::BasicBlock* block) {
+  SpvOp tailOp = block->tail()->opcode();
+  if (tailOp == SpvOpReturn || tailOp == SpvOpReturnValue) {
+    // bool shouldPredicate = return_flag_ != nullptr;
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Dead

File: source/opt/merge_return_pass.cpp [hunk 2]
```diff
@@ -15,28 +15,543 @@
...
+
+void MergeReturnPass::BranchToBlock(ir::BasicBlock* block, uint32_t target,
+                                    uint32_t structure_header) {
+  if (block->tail()->opcode() == SpvOpReturn ||
+      block->tail()->opcode() == SpvOpReturnValue) {
+    RecordReturned(block);
+    RecordReturnValue(block);
+  }
+
+  /*
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Dead

File: source/opt/merge_return_pass.cpp [hunk 3]
```diff
@@ -15,28 +15,543 @@
...
+      inst->AddOperand({SPV_OPERAND_TYPE_ID, {block->id()}});
+    }
+  });
+
+  if (structure_header != 0) {
+    auto predecessors = cfg()->preds(target);
+    if (predecessors.size() == 1) {
+      // Now that target will have more than 1 incoming edge, there may be some
+      // values that need a new OpPhi. We want to find values that are defined
+      // in blocks that use to dominate |target|, but do not anymore. If they
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: use -> used

File: source/opt/merge_return_pass.cpp [hunk 4]
```diff
@@ -15,28 +15,543 @@
...
+      // are used in a basic block that is dominated by |target|, then then a
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: double then

File: source/opt/merge_return_pass.cpp [hunk 5]
```diff
@@ -15,28 +15,543 @@
...
+      // new OpPhi is needed, and those uses replaced.
+      //
+      // Because the header of an OpLoopMerge or OpMerge structure dominates the
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: OpMerge -> OpSelectionMerge

File: source/opt/merge_return_pass.cpp [hunk 6]
```diff
@@ -15,28 +15,543 @@
...
+      // merge block (|target|), and the header is post-dominated by the merge
+      // block, the only blocks we need to consider are the immediate dominaters
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: dominaters -> dominators

File: source/opt/merge_return_pass.cpp [hunk 7]
```diff
@@ -15,28 +15,543 @@
...
+void MergeReturnPass::CreatePhiNodesForBlock(const ir::BasicBlock* block,
+                                             ir::BasicBlock* targetBlock,
+                                             uint32_t predecessor,
+                                             const ir::Instruction& inst) {
+  opt::DominatorAnalysis* dom_tree =
+      context()->GetDominatorAnalysis(block->GetParent(), *cfg());
+  if (inst.result_id() != 0) {
+    uint32_t result_of_phi = 0;
+    context()->get_def_use_mgr()->ForEachUse(
+        inst.result_id(),
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Change to &inst to save a lookup in the def map.

File: source/opt/merge_return_pass.cpp [hunk 8]
```diff
@@ -15,28 +15,543 @@
...
+    if (inst->opcode() == SpvOpPhi) {
+      uint32_t undefId = GetUndefForType(inst->type_id());
+      inst->AddOperand({SPV_OPERAND_TYPE_ID, {undefId}});
+      inst->AddOperand({SPV_OPERAND_TYPE_ID, {block->id()}});
+    }
+  });
+
+  if (structure_header != 0) {
+    auto predecessors = cfg()->preds(target);
+    if (predecessors.size() == 1) {
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Need to update the CFG for this to be reliable.

File: source/opt/merge_return_pass.cpp [hunk 9]
```diff
@@ -15,28 +15,543 @@
...
+                  {inst.result_id(), predecessor, undef_id, block->id()});
+              result_of_phi = new_phi->result_id();
+            }
+            user->SetOperand(index, {result_of_phi});
+          }
+        });
+  }
+}
+
+uint32_t MergeReturnPass::GetUndefForType(uint32_t type) {
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: I think this can be replaced with MemPass::Type2Undef

File: source/opt/merge_return_pass.cpp [hunk 10]
```diff
@@ -15,28 +15,543 @@
...
+}
+
+void MergeReturnPass::PredicateBlock(
+    ir::BasicBlock* block, ir::BasicBlock* tailBlock,
+    std::unordered_set<ir::BasicBlock*>* predicated) {
+  if (!RequiresPredication(block, tailBlock)) {
+    return;
+  }
+
+  std::cerr << "Predicating block: " << block->id() << std::endl;
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Remove

File: source/opt/merge_return_pass.cpp [hunk 11]
```diff
@@ -15,28 +15,543 @@
...
+  get_def_use_mgr()->AnalyzeInstUse(b->tailInst());
+  context()->set_instr_block(b->tailInst(), b);
+
+  // Within the new header we need the following:
+  // 1. Load of the return status flag
+  // 2. Declare the merge block
+  // 3. Branch to new merge (true) or old body (false)
+
+  // 1. Load of the return status flag
+  uint32_t boolId = get_module()->GetGlobalValue(SpvOpTypeBool);
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Use type manager here.

File: source/opt/merge_return_pass.cpp [hunk 12]
```diff
@@ -15,28 +15,543 @@
...
+
+  context()->InvalidateAnalyses(ir::IRContext::kAnalysisCFG);
+}
+
+void MergeReturnPass::RecordReturned(ir::BasicBlock* block) {
+  if (block->tail()->opcode() != SpvOpReturn &&
+      block->tail()->opcode() != SpvOpReturnValue)
+    return;
+
+  uint32_t boolId = get_module()->GetGlobalValue(SpvOpTypeBool);
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Use TypeManager::GetTypeInstruction() here.

File: source/opt/merge_return_pass.cpp [hunk 13]
```diff
@@ -15,28 +15,543 @@
...
+  if (boolId == 0) {
+    boolId = TakeNextId();
+    context()->AddType(
+        MakeUnique<ir::Instruction>(context(), SpvOpTypeBool, 0, boolId,
+                                    std::initializer_list<ir::Operand>{}));
+  }
+
+  if (!constant_true_) {
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Use constant manager.

File: source/opt/merge_return_pass.cpp [hunk 14]
```diff
@@ -15,28 +15,543 @@
...
+  insertIter.InsertBefore(std::move(returnValue));
+  return_value_ = &*function_->begin()->begin();
+  get_def_use_mgr()->AnalyzeInstDef(return_value_);
+}
+
+void MergeReturnPass::AddReturnFlag() {
+  if (return_flag_) return;
+
+  // TODO(alanbaker): fix this when we have a type and constant tables
+  uint32_t boolId = get_module()->GetGlobalValue(SpvOpTypeBool);
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Use type manager here.

File: source/opt/merge_return_pass.cpp [hunk 15]
```diff
@@ -15,28 +15,543 @@
...
+  if (boolId == 0) {
+    boolId = TakeNextId();
+    context()->AddType(
+        MakeUnique<ir::Instruction>(context(), SpvOpTypeBool, 0, boolId,
+                                    std::initializer_list<ir::Operand>{}));
+  }
+
+  uint32_t constFalseId = TakeNextId();
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Use constant manager.

File: source/opt/merge_return_pass.cpp [hunk 16]
```diff
@@ -47,33 +562,29 @@ std::vector<ir::BasicBlock*> MergeReturnPass::CollectReturnBlocks(
...
-
-  // Create the new basic block.
-  std::unique_ptr<ir::BasicBlock> returnBlock(
-      new ir::BasicBlock(std::move(returnLabel)));
-  function->AddBasicBlock(std::move(returnBlock));
-  ir::Function::iterator retBlockIter = --function->end();
+  CreateReturnBlock();
+  uint32_t returnId = final_return_block_->id();
+  auto retBlockIter = --function->end();
+  //// Create a label for the new return block
```
Comment (alan-baker) [inline source/opt/merge_return_pass.cpp]: Dead code

**Comment by s-perron** [comment]
I think the algorithm is looking good now, but it will be way to slow.  The cfg, def-use chains, and dominators are invalidated and rebuild in the one transformation multiple times.

There are still two things to get done.

- [ ] I still need to find the correct places and methods to update the analysies.
- [ ] I need to do more that just put in the spir-v for the test.  I need to actually test for something.

File: source/opt/basic_block.h [hunk 17]
```diff
@@ -103,6 +103,17 @@ class BasicBlock {
   // Returns true if the basic block has at least one successor.
   inline bool hasSuccessor() const { return ctail()->IsBranch(); }
 
+  // Returns a pointer to the last instruction in the basic block.
```
[the remark this answers is not in the document]
Comment (s-perron) [reply source/opt/basic_block.h]: Done.

**Comment by s-perron** [comment]
I think I have something that works.  I just need to add more comments, and make the code more readable in places.

File: source/opt/merge_return_pass.cpp [hunk 18]
```diff
@@ -15,118 +15,646 @@
...
+                                     uint32_t header_id) {
+  opt::DominatorAnalysis* dom_tree =
+      context()->GetDominatorAnalysis(function_, *cfg());
+  // Insert as a stopping point.  We do not have to add anything in the block or
+  // above because the header dominates |bb|.
+
+  ir::BasicBlock* current_bb = pred;
+  while (current_bb != nullptr && current_bb->id() != header_id) {
+    for (ir::Instruction& inst : *current_bb) {
+      CreatePhiNodesForInst(bb, current_bb->id(), inst);
```
[the remark this answers is not in the document]
Comment (s-perron) [reply source/opt/merge_return_pass.cpp]: right.  Added test, and fixed it.

File: source/opt/cfg.cpp [hunk 19]
```diff
@@ -139,5 +142,171 @@ void CFG::ComputePostOrderTraversal(BasicBlock* bb, vector<BasicBlock*>* order,
...
+  new_header->ForEachInst([new_header, context](ir::Instruction* inst) {
+    context->set_instr_block(inst, new_header);
+  });
+
+  // Adjust the OpPhi instructions as needed.
+  bb->ForEachPhiInst([latch_block, bb, new_header, context](Instruction* phi) {
+    std::vector<uint32_t> preheader_phi_ops;
+    std::vector<Operand> header_phi_ops;
+
+    // Idendify where the original inputs to original OpPhi belong: header or
```
Comment (dnovillo) [inline source/opt/cfg.cpp]: s/Idendify/Identify/

File: source/opt/cfg.h [hunk 20]
```diff
@@ -113,6 +113,21 @@ class CFG {
...
 
+  // Remove all edges that leave |bb|.
+  void RemoveSuccessorEdges(const ir::BasicBlock* bb) {
+    bb->ForEachSuccessorLabel(
+        [bb, this](uint32_t succ_id) { RemoveEdge(bb->id(), succ_id); });
+  }
+
+  // Divides |block| into two basic blocks.  The first block will have the same
+  // id as |block| and will become a preheader for the loop.  The other block
+  // becomes is a new block that will be the new loop header.
```
Comment (dnovillo) [inline source/opt/cfg.h]: s/becomes //

File: source/opt/dominator_tree.cpp [hunk 21]
```diff
@@ -211,6 +209,13 @@ void BasicBlockSuccessorHelper<BBType>::CreateSuccessorMap(
         predecessors_[succ].push_back(&bb);
       });
     }
+    /*
+    for (BasicBlock& bb : f) {
+      if (predecessors_[&bb].empty()) {
+        predecessors_[&bb].push_back(const_cast<BasicBlock*>(dummy_start_node));
+        successors_[dummy_start_node].push_back(&bb);
+      }
+    } */
```
Comment (dnovillo) [inline source/opt/dominator_tree.cpp]: This seems like stale code.

File: source/opt/ir_context.cpp [hunk 22]
```diff
@@ -577,5 +585,56 @@ opt::PostDominatorAnalysis* IRContext::GetPostDominatorAnalysis(
...
+      }
+
+      for (size_t i = 0; i < real.size() && same; i++) {
+        if (preds[i] != real[i]) {
+          same = false;
+        }
+      }
+
+      if (!same) {
+        std::cerr << "Predecessors for " << bb.id() << " are different:\n";
```
Comment (dnovillo) [inline source/opt/ir_context.cpp]: I think we should stop execution here.

File: source/opt/merge_return_pass.h [hunk 23]
```diff
@@ -17,36 +17,288 @@
...
+ * In the above (simple) example, the return originally in |2| is passed through
+ * the merge. That merge is predicated such that the old body of the block is
+ * the else branch. The branch condition is based on the value of the "has
+ * returned" variable. In more complicated examples (blocks between |1| and
+ * |3|), the SSA would need to fixed up due the newly reconvergent path at the
+ * merge for |1|. Assuming |3| originally was also a return block, the old body
+ * of |3| should also store the return value for that case. The return value in
+ * |4| just requires loading the return value variable.
+ *
+ ******************************************************************************/
```
Comment (dnovillo) [inline source/opt/merge_return_pass.h]: Thanks for this!

Comment (s-perron) [reply source/opt/merge_return_pass.h]: That was Alan that wrote that.

File: source/opt/ir_context.cpp (hunk 22 above)
Comment (s-perron) [reply source/opt/ir_context.cpp]: The extra output was very useful.  You can see what is missing.  I don't want to have to rewrite it every time I need it.

**Comment by dnovillo** [comment]
On Wed, Mar 21, 2018, 09:51 Steven Perron, <notifications@github.com> wrote:

> *@s-perron* commented on this pull request.
> ------------------------------
>
> In source/opt/ir_context.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/1380#discussion_r176156996>
> :
>
> > +      std::sort(preds.begin(), preds.end());
> +      std::sort(real.begin(), real.end());
> +
> +      bool same = true;
> +      if (preds.size() != real.size()) {
> +        same = false;
> +      }
> +
> +      for (size_t i = 0; i < real.size() && same; i++) {
> +        if (preds[i] != real[i]) {
> +          same = false;
> +        }
> +      }
> +
> +      if (!same) {
> +        std::cerr << "Predecessors for " << bb.id() << " are different:\n";
>
> The extra output was very useful. You can see what is missing. I don't
> want to have to rewrite it every time I need it.
>
I suggest ifdedfing around this debugging code. I have a bunch of that in
the ssa rewriter. We can convert those to debug logs when the debug logger
is implemented.

>

**Comment by s-perron** [comment]
That code is already ifdef at a higher level.  The whole function is a debug thing, and we assert if the return value if false.

**Comment by dnovillo** [comment]
On Wed, Mar 21, 2018, 10:02 Steven Perron, <notifications@github.com> wrote:

> That code is already ifdef at a higher level. The whole function is a
> debug thing, and we assert if the return value if false.
>

Oh, ok. In that case, never mind me :)

**Pull request merged**